<center><h1><font size="6">DataDive Event Notebook</font></h1>
<h2>Author: Steven Yan</h2></center>



In [1]:
# for Google Colab only
# from google.colab import drive
# drive.mount('/content/drive')

# <a id='1'>Importing Packages</a>

In [2]:
import pandas as pd 
import numpy as np
# ! pip install geopandas
import geopandas as gpd 

# Set maximum number of rows and columns that can be viewed
pd.set_option("display.max_columns", 100)
pd.set_option('display.max_rows', 100)


# <a id='2'>Importing Databases</a>

<a href='#0'>Return to Top</a>

Since there seems to be overlap in the terminology, it's best to take a look at how the various databases uses the same nomenclature as the Census Bureau. Generally, state FIPS codes are usually two digits, and county-level FIPS codes are either three or five digits since the state and county FIPS codes can be concatenated to make the five-digit code.


## ACS Database

In [4]:
# Importing ACS database
# data = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/National/acs_2019.csv.gz', compression ='gzip')
data = pd.read_csv('data/ACS/acs_2019.csv.gz', compression ='gzip')
data.head(20)

state  county  tract       geoid  f_broadband  f_computer   f_ba  f_black  \
0       1       1  20100  1001020100        0.740       0.815  0.271    0.113   
1       1       1  20200  1001020200        0.669       0.747  0.171    0.582   
2       1       1  20300  1001020300        0.804       0.879  0.184    0.260   
3       1       1  20400  1001020400        0.856       0.926  0.348    0.070   
4       1       1  20500  1001020500        0.890       0.947  0.424    0.192   
5       1       1  20600  1001020600        0.837       0.943  0.216    0.204   
6       1       1  20700  1001020700        0.766       0.862  0.103    0.249   
7       1       1  20801  1001020801        0.916       0.958  0.353    0.139   
8       1       1  20802  1001020802        0.851       0.903  0.257    0.145   
9       1       1  20900  1001020900        0.712       0.876  0.227    0.102   
10      1       1  21000  1001021000        0.748       0.899  0.204    0.204   
11      1       1  21100  1001021100        0.544       0.710  0.116    0.521   
12      1       3  10100  1003010100        0.730       0.811  0.116    0.110   
13      1       3  10200  1003010200        0.766       0.828  0.097    0.080   
14      1       3  10300  1003010300        0.755       0.819  0.235    0.157   
15      1       3  10400  1003010400        0.713       0.873  0.207    0.087   
16      1       3  10500  1003010500        0.817       0.929  0.208    0.122   
17      1       3  10600  1003010600        0.680       0.774  0.095    0.629   
18      1       3  10701  1003010701        0.891       0.934  0.496    0.019   
19      1       3  10703  1003010703        0.936       0.981  0.448    0.097   

    f_hispanic  log_mhi      mhi  population  households  broadband  computer  \
0        0.013    11.01  60208.0      1993.0       709.0       74.0      81.5   
1        0.015    10.69  43958.0      1959.0       688.0       66.9      74.7   
2        0.066    10.92  55345.0      3507.0      1360.0       80.4      87.9   
3        0.011    11.00  59663.0      3878.0      1675.0       85.6      92.6   
4        0.044    11.10  66108.0     10596.0      4483.0       89.0      94.7   
5        0.045    10.72  45096.0      3668.0      1340.0       83.7      94.3   
6        0.070    10.72  45444.0      3586.0      1338.0       76.6      86.2   
7        0.013    11.29  80174.0      3196.0      1110.0       91.6      95.8   
8        0.019    11.10  65878.0     11142.0      4098.0       85.1      90.3   
9        0.004    11.11  66726.0      6143.0      2295.0       71.2      87.6   
10       0.012    10.80  48981.0      2527.0       986.0       74.8      89.9   
11       0.011    10.59  39680.0      3185.0      1315.0       54.4      71.0   
12       0.000    10.74  45990.0      3433.0      1395.0       73.0      81.1   
13       0.000    10.54  37770.0      3158.0      1246.0       76.6      82.8   
14       0.018    10.97  58056.0      8284.0      2756.0       75.5      81.9   
15       0.014    10.88  53250.0      5181.0      1849.0       71.3      87.3   
16       0.000    10.86  52083.0      5463.0      1870.0       81.7      92.9   
17       0.053    10.35  31229.0      3730.0      1122.0       68.0      77.4   
18       0.050    11.35  85088.0      8756.0      3335.0       89.1      93.4   
19       0.023    11.46  94611.0     19962.0      6874.0       93.6      98.1   

    black  hispanic    mhi.1    ba  den_computers  n_computer  n_broadband  \
0    11.3       1.3  60208.0  27.1          709.0       578.0        525.0   
1    58.2       1.5  43958.0  17.1          688.0       514.0        460.0   
2    26.0       6.6  55345.0  18.4         1360.0      1196.0       1093.0   
3     7.0       1.1  59663.0  34.8         1675.0      1551.0       1434.0   
4    19.2       4.4  66108.0  42.4         4483.0      4246.0       3991.0   
5    20.4       4.5  45096.0  21.6         1340.0      1264.0       1121.0   
6    24.9       7.0  45444.0  10.3         1338.0     

In [5]:
# Get the counts of the values in tract column
len(data.tract.value_counts())

23816

In [6]:
# Get the counts of the values in geoid column
len(data.geoid.value_counts())

73056

There are substantially more geoids than unique tract values, so different regions may have the same tract value since geoids are distinct and non-overlapping regions.

In [6]:
data.columns

Index(['state', 'county', 'tract', 'geoid', 'f_broadband', 'f_computer',
       'f_ba', 'f_black', 'f_hispanic', 'log_mhi', 'mhi', 'population',
       'households', 'broadband', 'computer', 'black', 'hispanic', 'mhi.1',
       'ba', 'den_computers', 'n_computer', 'n_broadband', 'den_black',
       'n_black', 'den_hispanic', 'n_hispanic', 'den_ba', 'n_ba',
       'nhh_computer', 'nhh_broadband', 'nhh_computer_any_internet',
       'nhh_computer_and_dialup', 'nhh_computer_and_broadband',
       'nhh_computer_no_internet', 'nhh_no_computer', 'den_age', 'n_children',
       'n_children_computer', 'n_children_computer_and_dialup',
       'n_children_computer_and_broadband', 'n_children_computer_no_internet',
       'n_children_no_computer'],
      dtype='object')

In [7]:
# shorten column names to less than 10 characters
data.columns = ['state', 'county', 'tract', 'geoid', 'f_bb', 'f_comp', 'f_ba', 'f_black', 'f_hisp', 'log_mhi', 'mhi', 'population', 'hh', 'bb', 'comp', 'black', 'hisp', 'mhi.1', 'ba', 'den_comps', 'n_comp', 'n_bb', 'den_black', 'n_black', 'den_hisp', 'n_hisp', 'den_ba', 'n_ba',
       'nhh_c', 'nhh_b', 'nhh_c_a_i',
       'nhh_c_a_d', 'nhh_c_a_b',
       'nhh_c_n_i', 'nhh_n_c', 'den_age', 'n_ch',
       'n_ch_c', 'n_ch_c_n_d',
       'n_ch_c_a_b', 'n_ch_c_n_i',
       'n_ch_n_c']

Notes:

- `den_computers` = `households` = `nhh_computer` + `nhh_no_computer`
-  can eliminate `state`, `county`, `tract` since `state` + `county` + `tract` = `geoid`
- `f_variable` is variable / 100, i.e. `f_broadband` = `broadband` / 100
- `den_hispanic` = `den_black` = `population`
- `n_computer` = `nhh_computer` = `nhh_computer_any_internet`
- `n_broadband` = `nhh_broadband`-. `n_children` = `n_children_computer` = `n_children_computer_and dialup` + `n_cildren_computer_and_broadband` + `n_children_computer_no_internet` + `n_children_no_computer`
- eliminate mhi.1
- `nhh_no_broadband` = `households` - `nhh_broadband`
- `n_children_no_computer_and_broadband` = `n_children` - `n_children_computer_and_broadband`

I haven't been able to entirely figure out most of the columns mean.  n is either no or number.  nhh is no household?

# <a id='2'>State-level databases</a>

<a href='#0'>Return to Top</a>

## Geodata


In [7]:
# import shapefile into dataframe
# ny_shp = gpd.read_file("/content/drive/MyDrive/Datasets/Broadband/NY/ny_spdf/ny_spdf.shp")
ny_shp = gpd.read_file("data/NY/ny_spdf/ny_spdf.shp")
ny_shp.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID     NAME LSAD  \
0      36      071  000502  1400000US36071000502  36071000502     5.02   CT   
1      36      103  135208  1400000US36103135208  36103135208  1352.08   CT   
2      36      047  057800  1400000US36047057800  36047057800      578   CT   
3      36      047  058900  1400000US36047058900  36047058900      589   CT   
4      36      055  013204  1400000US36055013204  36055013204   132.04   CT   

        ALAND    AWATER                                           geometry  
0    967431.0  969216.0  POLYGON ((-74.02226 41.49281, -74.02180 41.496...  
1   2287077.0       0.0  POLYGON ((-73.28263 40.83063, -73.28157 40.832...  
2    172233.0       0.0  POLYGON ((-73.95398 40.60140, -73.95304 40.601...  
3    424025.0   38353.0  POLYGON ((-73.94605 40.72926, -73.94419 40.729...  
4  28207247.0   93541.0  POLYGON ((-77.66835 43.02829, -77.66806 43.029...

In [8]:
# get some information about GEOID
print("Number of unique values: ", ny_shp.GEOID.nunique())
print("Minimum value: ", ny_shp.GEOID.min())
print("Maximum value: ", ny_shp.GEOID.max())

Number of unique values:  4906
Minimum value:  36001000100
Maximum value:  36123150500


In [10]:
# get some information about TRACTCE
print("Number of unique values: ", ny_shp.TRACTCE.nunique())
print("Minimum value: ", ny_shp.TRACTCE.min())
print("Maximum value: ", ny_shp.TRACTCE.max())

Number of unique values:  2697
Minimum value:  000100
Maximum value:  990100


In [18]:
# sort dataframe by GEOID values
ny_shape = ny_shp.sort_values(by=['GEOID'])
ny_shape.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD  \
539       36      001  000100  1400000US36001000100  36001000100     1   CT   
2748      36      001  000200  1400000US36001000200  36001000200     2   CT   
3007      36      001  000300  1400000US36001000300  36001000300     3   CT   
3515      36      001  000401  1400000US36001000401  36001000401  4.01   CT   
2751      36      001  000403  1400000US36001000403  36001000403  4.03   CT   

          ALAND    AWATER                                           geometry  
539   2367456.0  245021.0  POLYGON ((-73.74506 42.67228, -73.74374 42.677...  
2748  2065161.0       0.0  POLYGON ((-73.75993 42.65997, -73.76156 42.660...  
3007  5758852.0   55326.0  POLYGON ((-73.83021 42.69418, -73.82785 42.696...  
3515  9065815.0   86327.0  POLYGON ((-73.89421 42.70977, -73.89668 42.711...  
2751  3138700.0       0.0  POLYGON ((-73.82108 42.67857, -73.81916 42.680...

## Synopsis of NY Shapefile:

There are 4906 rows in the shapefile, as well 
as 4906 unique rows of GEOID's. The state FIPS code is 36, and there are 62 county FIPS code in NY.  The AFFGEOID seems to be the GEOID affixed with 1400000US in front of each GEOID.  There are values given for land area, water area, and most importantly the geometry of the tract boundaries.

Originally, I thought that TRACTCEs needed to be concatentated together as they each represented a unique value, but instead they represent a 6-digit tract FIPS code.  

Each GEOID consists of the state FIPS code, its county FIPS code, and then the tract FIPS code, which apparently in itself is not unique and can be repeated, so 2 + 3 + 6 digits in total for the three components or an 11-digit tract FIPS code that is unique for every tract.

## State-level ACS Database

In [33]:
# import database
# ny_1 = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/NY/acs_2019_NY.csv')
ny_acs = pd.read_csv('data/NY/acs_2019_NY.csv')
ny_acs.head(20)

state  county  tract        geoid  f_broadband  f_computer   f_ba  \
0      36       1    100  36001000100        0.836       0.896  0.137   
1      36       1    200  36001000200        0.767       0.903  0.225   
2      36       1    300  36001000300        0.819       0.881  0.289   
3      36       1    401  36001000401        0.837       0.857  0.476   
4      36       1    403  36001000403        0.817       0.861  0.591   
5      36       1    404  36001000404        1.000       1.000  0.073   
6      36       1    501  36001000501        0.718       0.811  0.228   
7      36       1    502  36001000502        0.806       0.945  0.621   
8      36       1    600  36001000600        0.747       0.912  0.183   
9      36       1    700  36001000700        0.751       0.834  0.167   
10     36       1    800  36001000800        0.672       0.842  0.112   
11     36       1   1100  36001001100        0.534       0.673  0.228   
12     36       1   1400  36001001400        0.840       0.933  0.609   
13     36       1   1500  36001001500        0.740       0.890  0.539   
14     36       1   1600  36001001600        0.897       0.915  0.599   
15     36       1   1700  36001001700        0.814       0.898  0.539   
16     36       1   1801  36001001801        0.825       0.904  0.550   
17     36       1   1802  36001001802        0.838       0.852  0.408   
18     36       1   1901  36001001901        0.919       0.951  0.661   
19     36       1   1902  36001001902        0.868       0.904  0.431   

    f_black  f_hispanic  log_mhi          mhi  population  households  \
0     0.740       0.128    10.39      32389.0      2035.0       800.0   
1     0.845       0.043    10.23      27714.0      4793.0      2213.0   
2     0.435       0.121    10.72      45272.0      6147.0      2362.0   
3     0.076       0.046    11.22      74274.0      2362.0       997.0   
4     0.129       0.058    11.22      74426.0      4253.0      2186.0   
5     0.225       0.190      NaN -666666666.0      5082.0        15.0   
6     0.439       0.226    10.38      32165.0      3523.0      1433.0   
7     0.201       0.106    11.01      60263.0      3785.0       942.0   
8     0.363       0.137    10.01      22298.0      3663.0      1408.0   
9     0.807       0.099    10.51      36639.0      3885.0      1485.0   
10    0.799       0.211    10.16      25872.0      2192.0       842.0   
11    0.375       0.097     9.89      19766.0      1573.0       893.0   
12    0.206       0.070    10.81      49602.0      3855.0      2816.0   
13    0.123       0.065    10.39      32500.0      4775.0      2254.0   
14    0.132       0.050    10.83      50417.0      4151.0      1896.0   
15    0.173       0.041    11.07      63913.0      4318.0      1964.0   
16    0.081       0.057    11.41      90043.0      7152.0      3494.0   
17    0.100       0.034    11.16      69918.0      4068.0      1849.0   
18    0.084       0.025    11.32      82713.0      2714.0      1215.0   
19    0.140       0.040    11.30      81081.0      2293.0      1036.0   

    broadband  computer  black  hispanic        mhi.1    ba  den_computers  \
0        83.6      89.6   74.0      12.8      32389.0  13.7          800.0   
1        76.7      90.3   84.5       4.3      27714.0  22.5         2213.0   
2        81.9      88.1   43.5      12.1      45272.0  28.9         2362.0   
3        83.7      85.7    7.6       4.6      74274.0  47.6          997.0   
4        81.7      86.1   12.9       5.8      74426.0  59.1         2186.0   
5       100.0     100.0   22.5      19.0 -666666666.0   7.3           15.0   
6        71.8      81.1   43.9      22.6      32165.0  22.8         1433.0   
7        80.6      94.5   20.1      10.6      60263.0  62.1          942.0   
8        74.7      91.2   36.3      13.7      22298.0  18.3         1408.0   
9        75.1      83.4   80.7       9.9      36639.0  16.7         1485.0   
10       67.2      84.2   79.9      21.1      25872.0  11.2          842.0   


In [34]:
ny_acs.drop(columns=['state', 'county', 'tract', 'broadband', 'computer', 'black', 'hispanic', 'mhi.1', 'den_computers', 'n_black', 'n_hispanic', 'den_black', 'den_hispanic', 'state_lkp', 'den_age', 'n_computer', 'n_broadband'], inplace=True)
ny_acs

geoid  f_broadband  f_computer   f_ba  f_black  f_hispanic  \
0     36001000100        0.836       0.896  0.137    0.740       0.128   
1     36001000200        0.767       0.903  0.225    0.845       0.043   
2     36001000300        0.819       0.881  0.289    0.435       0.121   
3     36001000401        0.837       0.857  0.476    0.076       0.046   
4     36001000403        0.817       0.861  0.591    0.129       0.058   
...           ...          ...         ...    ...      ...         ...   
4913  36123150100        0.780       0.833  0.270    0.012       0.008   
4914  36123150200        0.688       0.743  0.189    0.023       0.038   
4915  36123150300        0.754       0.849  0.238    0.014       0.020   
4916  36123150400        0.824       0.842  0.338    0.010       0.029   
4917  36123150500        0.744       0.792  0.257    0.017       0.020   

      log_mhi      mhi  population  households    ba  den_ba    n_ba  \
0       10.39  32389.0      2035.0       800.0  13.7  1283.0   176.0   
1       10.23  27714.0      4793.0      2213.0  22.5  2710.0   609.0   
2       10.72  45272.0      6147.0      2362.0  28.9  3725.0  1078.0   
3       11.22  74274.0      2362.0       997.0  47.6  2169.0  1032.0   
4       11.22  74426.0      4253.0      2186.0  59.1  3220.0  1904.0   
...       ...      ...         ...         ...   ...     ...     ...   
4913    10.90  54330.0      6288.0      2293.0  27.0  4283.0  1155.0   
4914    10.77  47586.0      5440.0      1874.0  18.9  3661.0   692.0   
4915    10.93  55694.0      5092.0      1858.0  23.8  3438.0   818.0   
4916    11.18  71774.0      3763.0      1196.0  33.8  2176.0   735.0   
4917    10.92  55417.0      4428.0      1698.0  25.7  3115.0   801.0   

      nhh_computer  nhh_broadband  nhh_computer_any_internet  \
0              717            669                        717   
1             1999           1697                       1999   
2             2081           1935                       2081   
3              854            834                        854   
4             1882           1786                       1882   
...            ...            ...                        ...   
4913          1910           1789                       1910   
4914          1392           1290                       1392   
4915          1577           1401                       1577   
4916          1007            986                       1007   
4917          1345           1263                       1345   

      nhh_computer_and_dialup  nhh_computer_and_broadband  \
0                           0                         669   
1                           0                        1685   
2                          11                        1917   
3                           0                         826   
4                           0                        1733   
...                       ...                         ...   
4913                        8                        1774   
4914                        6                        1272   
4915                       13                        1386   
4916                        0                         951   
4917                       15                        1248   

      nhh_computer_no_internet  nhh_no_computer  n_children  \
0                           48               83         532   
1                          314              214        1274   
2                          153              281        1729   
3                           28              143         171   
4                          149              304         462   
...                        ...              ...         ...   
4913                       128              383        1533   
4914                       114              482        1219   
4915                       178              281        1213   
4916                        56              189         592   
4917                        82              353         945   


In [35]:
# get some information about geoid
print("Number of unique values: ", ny_acs.geoid.nunique())
print("Minimum value: ", ny_acs.geoid.min())
print("Maximum value: ", ny_acs.geoid.max())


Number of unique values:  4918
Minimum value:  36001000100
Maximum value:  36123150500


There are more Geoids in the ACS dataset than in the shapefile.  4918 vs. 4906.

In [36]:
ny_acs['geoid'] = ny_acs['geoid'].apply(lambda x: str(x))

In [37]:
joined_df = ny_shape.merge(ny_acs, how='left', left_on='GEOID', right_on='geoid')
joined_df.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD  \
0      36      001  000100  1400000US36001000100  36001000100     1   CT   
1      36      001  000200  1400000US36001000200  36001000200     2   CT   
2      36      001  000300  1400000US36001000300  36001000300     3   CT   
3      36      001  000401  1400000US36001000401  36001000401  4.01   CT   
4      36      001  000403  1400000US36001000403  36001000403  4.03   CT   

       ALAND    AWATER                                           geometry  \
0  2367456.0  245021.0  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  2065161.0       0.0  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  5758852.0   55326.0  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  9065815.0   86327.0  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  3138700.0       0.0  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

         geoid  f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi  \
0  36001000100        0.836       0.896  0.137    0.740       0.128    10.39   
1  36001000200        0.767       0.903  0.225    0.845       0.043    10.23   
2  36001000300        0.819       0.881  0.289    0.435       0.121    10.72   
3  36001000401        0.837       0.857  0.476    0.076       0.046    11.22   
4  36001000403        0.817       0.861  0.591    0.129       0.058    11.22   

       mhi  population  households    ba  den_ba    n_ba  nhh_computer  \
0  32389.0      2035.0       800.0  13.7  1283.0   176.0           717   
1  27714.0      4793.0      2213.0  22.5  2710.0   609.0          1999   
2  45272.0      6147.0      2362.0  28.9  3725.0  1078.0          2081   
3  74274.0      2362.0       997.0  47.6  2169.0  1032.0           854   
4  74426.0      4253.0      2186.0  59.1  3220.0  1904.0          1882   

   nhh_broadband  nhh_computer_any_internet  nhh_computer_and_dialup  \
0            669                        717                        0   
1           1697                       1999                        0   
2           1935                       2081                       11   
3            834                        854                        0   
4           1786                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   n_children  n_children_computer  n_children_computer_and_dialup  \
0         532                  520                               0   
1        1274                 1221                               0   
2        1729                 1717                               0   
3         171                  171                               0   
4         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer  
0                      12  
1                      53  
2                      12  
3                       0  
4                       0

In [38]:
joined_df.drop(columns=['STATEFP', 'COUNTYFP', 'TRACTCE', 'AFFGEOID', 'NAME', 'LSAD', 'ALAND', 'AWATER'], inplace=True)

In [39]:
joined_df.head()

GEOID                                           geometry  \
0  36001000100  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  36001000200  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  36001000300  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  36001000401  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  36001000403  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

         geoid  f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi  \
0  36001000100        0.836       0.896  0.137    0.740       0.128    10.39   
1  36001000200        0.767       0.903  0.225    0.845       0.043    10.23   
2  36001000300        0.819       0.881  0.289    0.435       0.121    10.72   
3  36001000401        0.837       0.857  0.476    0.076       0.046    11.22   
4  36001000403        0.817       0.861  0.591    0.129       0.058    11.22   

       mhi  population  households    ba  den_ba    n_ba  nhh_computer  \
0  32389.0      2035.0       800.0  13.7  1283.0   176.0           717   
1  27714.0      4793.0      2213.0  22.5  2710.0   609.0          1999   
2  45272.0      6147.0      2362.0  28.9  3725.0  1078.0          2081   
3  74274.0      2362.0       997.0  47.6  2169.0  1032.0           854   
4  74426.0      4253.0      2186.0  59.1  3220.0  1904.0          1882   

   nhh_broadband  nhh_computer_any_internet  nhh_computer_and_dialup  \
0            669                        717                        0   
1           1697                       1999                        0   
2           1935                       2081                       11   
3            834                        854                        0   
4           1786                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   n_children  n_children_computer  n_children_computer_and_dialup  \
0         532                  520                               0   
1        1274                 1221                               0   
2        1729                 1717                               0   
3         171                  171                               0   
4         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer  
0                      12  
1                      53  
2                      12  
3                       0  
4                       0

In [40]:
type(joined_df)

geopandas.geodataframe.GeoDataFrame

In [26]:
joined_df.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Int64Index: 4906 entries, 0 to 4905
Data columns (total 40 columns):
 #   Column                             Non-Null Count  Dtype   
---  ------                             --------------  -----   
 0   GEOID                              4906 non-null   object  
 1   geometry                           4906 non-null   geometry
 2   f_broadband                        4906 non-null   float64 
 3   f_computer                         4906 non-null   float64 
 4   f_ba                               4906 non-null   float64 
 5   f_black                            4906 non-null   float64 
 6   f_hispanic                         4906 non-null   float64 
 7   log_mhi                            4785 non-null   float64 
 8   mhi                                4906 non-null   float64 
 9   population                         4906 non-null   float64 
 10  households                         4906 non-null   float64 
 11  broadband                          

There are currrently 4906 instances corresponding each of the census tracts in the shapefile.

## FCC Database

The columns are as follows:

- `tract`: 11-digit tract FIPS code
- `max_dn`: average maximum advertised downstream speed offered by provider in census tract by block
- `max_up`: average maximum advertised upstream speed offered by provider in census tract by block
- `dn10`:	average count of providers offering advertised downstream greater than 10 by block
- `dn100`: average count of providers offering advertised downstream greater than 100 by block
- `dn250`: average count of providers offering advertised downstream greater than 250 by block
- `fiber_100u`: average count of providers offiering fiber by block


In [42]:
# importing FCC database
# ny_fcc = pd.read_csv('/content/drive/MyDrive/Datasets/DataKind/NY/fcc_477_census_tract_NY.csv')
ny_fcc = pd.read_csv('data/NY/fcc_477_census_tract_NY.csv')

ny_fcc.head()

tract      max_dn     max_up      dn10     dn100     dn250  \
0  36001000100  711.250000  50.722222  1.875000  0.777778  0.777778   
1  36001000200  940.000000  35.000000  2.529412  1.011765  1.000000   
2  36001000300  928.040541  98.195946  2.290541  1.060811  1.060811   
3  36001000401  928.975904  65.168675  2.156627  1.024096  1.024096   
4  36001000403  940.000000  43.203883  2.310680  1.009709  1.009709   

   fiber_100u  state  
0    0.027778     36  
1    0.000000     36  
2    0.074324     36  
3    0.036145     36  
4    0.009709     36

In [44]:
# check for any null values in dataframe
ny_fcc.isna().sum()

tract         0
max_dn        0
max_up        0
dn10          0
dn100         0
dn250         0
fiber_100u    0
state         0
dtype: int64

In [45]:
# show value counts of tract column
len(ny_fcc.tract.value_counts())

4900

In [46]:
# change tract column to string type for joining
ny_fcc['tract'] = ny_fcc['tract'].apply(lambda x: str(x))
type(ny_fcc['tract'][0])

str

In [47]:
# merge two database on the unique 11-digit FIPS code
joined_2 = joined_df.merge(ny_fcc, how='left', left_on='GEOID', right_on='tract')
joined_2.head()

GEOID                                           geometry  \
0  36001000100  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  36001000200  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  36001000300  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  36001000401  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  36001000403  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

         geoid  f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi  \
0  36001000100        0.836       0.896  0.137    0.740       0.128    10.39   
1  36001000200        0.767       0.903  0.225    0.845       0.043    10.23   
2  36001000300        0.819       0.881  0.289    0.435       0.121    10.72   
3  36001000401        0.837       0.857  0.476    0.076       0.046    11.22   
4  36001000403        0.817       0.861  0.591    0.129       0.058    11.22   

       mhi  population  households    ba  den_ba    n_ba  nhh_computer  \
0  32389.0      2035.0       800.0  13.7  1283.0   176.0           717   
1  27714.0      4793.0      2213.0  22.5  2710.0   609.0          1999   
2  45272.0      6147.0      2362.0  28.9  3725.0  1078.0          2081   
3  74274.0      2362.0       997.0  47.6  2169.0  1032.0           854   
4  74426.0      4253.0      2186.0  59.1  3220.0  1904.0          1882   

   nhh_broadband  nhh_computer_any_internet  nhh_computer_and_dialup  \
0            669                        717                        0   
1           1697                       1999                        0   
2           1935                       2081                       11   
3            834                        854                        0   
4           1786                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   n_children  n_children_computer  n_children_computer_and_dialup  \
0         532                  520                               0   
1        1274                 1221                               0   
2        1729                 1717                               0   
3         171                  171                               0   
4         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer        tract      max_dn     max_up      dn10  \
0                      12  36001000100  711.250000  50.722222  1.875000   
1                      53  36001000200  940.000000  35.000000  2.529412   
2                      12  36001000300  928.040541  98.195946  2.290541   
3                       0  36001000401  928.975904  65.168675  2.156627   
4                       0  36001000403  940.000000  43.203883  2.310680   

      dn100     dn250  fiber_100u  state  
0  0.777778  0.777778    0.027778   36.0  
1  1.011765  1.000000    0.000000   36.0  
2  1.060811  1.060811    0.074324   36.0  
3  1.024096  1.024096    0.036145   36.0  
4  1.009709  1.009709    0.009709   36.0

In [48]:
joined_2.drop(columns=['tract', 'geoid'], inplace=True)

In [55]:
joined_2.head()

GEOID                                           geometry  \
0  36001000100  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  36001000200  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  36001000300  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  36001000401  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  36001000403  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

   f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi      mhi  \
0        0.836       0.896  0.137    0.740       0.128    10.39  32389.0   
1        0.767       0.903  0.225    0.845       0.043    10.23  27714.0   
2        0.819       0.881  0.289    0.435       0.121    10.72  45272.0   
3        0.837       0.857  0.476    0.076       0.046    11.22  74274.0   
4        0.817       0.861  0.591    0.129       0.058    11.22  74426.0   

   population  households    ba  den_ba    n_ba  nhh_computer  nhh_broadband  \
0      2035.0       800.0  13.7  1283.0   176.0           717            669   
1      4793.0      2213.0  22.5  2710.0   609.0          1999           1697   
2      6147.0      2362.0  28.9  3725.0  1078.0          2081           1935   
3      2362.0       997.0  47.6  2169.0  1032.0           854            834   
4      4253.0      2186.0  59.1  3220.0  1904.0          1882           1786   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                        717                        0   
1                       1999                        0   
2                       2081                       11   
3                        854                        0   
4                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   n_children  n_children_computer  n_children_computer_and_dialup  \
0         532                  520                               0   
1        1274                 1221                               0   
2        1729                 1717                               0   
3         171                  171                               0   
4         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer      max_dn     max_up      dn10     dn100  \
0                      12  711.250000  50.722222  1.875000  0.777778   
1                      53  940.000000  35.000000  2.529412  1.011765   
2                      12  928.040541  98.195946  2.290541  1.060811   
3                       0  928.975904  65.168675  2.156627  1.024096   
4                       0  940.000000  43.203883  2.310680  1.009709   

      dn250  fiber_100u  state  
0  0.777778    0.027778   36.0  
1  1.000000    0.000000   36.0  
2  1.060811    0.074324   36.0  
3  1.024096    0.036145   36.0  
4  1.009709    0.009709   36.0

## Synopsis of FCC Database

FCC has 4900 rows of data, and the shapefile had 4906 rows of each of the tracts, and every row has a unique tract value. Here `tract` is synonymous with `GEOID`. It would seem we are missing data for a few GEOIDs, but that remains to be seen how our analysis will be affected.

## Ookla Database

The column keys are:
- `avg_d_kbps`, `int`, average download speed of all tests performed in the tile, represented in kilobits per second.
- `avg_u_kbps`, `int`, 	average upload speed of all tests performed in the tile, represented in kilobits per second.
- `avg_lat_ms`, `int`, average latency of all tests performed in the tile, represented in milliseconds
- `tests`, `int`, number of tests taken in the tile.
- `devices`, `int`, number of unique devices contributing tests in the tile.
- `quadkey`: quadkey representing the tile.



In [49]:
# import Ookla database
ny_ookla = pd.read_csv('data/NY/ookla_combined_ny.csv')
# ny_3 = pd.read_csv('/content/drive/MyDrive/Datasets/Broadband/NY/ookla_combined_ny.csv')
ny_ookla.head()

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices   type  \
0  302332203102320      208526        7394          14     50        2  fixed   
1  302231312100203      153742       97895           6    134       28  fixed   
2  320101103112201      202082       55540          15    222       56  fixed   
3  302322000030131      176743       14699          20     18        3  fixed   
4  302233111212211      370517      311427           4      2        1  fixed   

  quarter  year  STATEFP  COUNTYFP  TRACTCE              AFFGEOID  \
0      Q1  2021       36       103   170202  1400000US36103170202   
1      Q1  2021       36        63    24502  1400000US36063024502   
2      Q1  2021       36        47    92400  1400000US36047092400   
3      Q1  2021       36        29    14702  1400000US36029014702   
4      Q1  2021       36        29     7302  1400000US36029007302   

         GEOID     NAME LSAD     ALAND     AWATER  \
0  36103170202  1702.02   CT  41367468  100616534   
1  36063024502   245.02   CT  77726398   10786358   
2  36047092400   924.00   CT    251687          0   
3  36029014702   147.02   CT  31997799     308205   
4  36029007302    73.02   CT  36949030    5791136   

                                                tile  
0  list(c(-72.57568359375, -72.5701904296875, -72...  
1  list(c(-79.2718505859375, -79.266357421875, -7...  
2  list(c(-73.9105224609375, -73.905029296875, -7...  
3  list(c(-78.6236572265625, -78.6181640625, -78....  
4  list(c(-78.9971923828125, -78.99169921875, -78...

In [50]:
# get general attributes of dataframe
ny_ookla.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 444771 entries, 0 to 444770
Data columns (total 19 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   quadkey     444771 non-null  int64  
 1   avg_d_kbps  444771 non-null  int64  
 2   avg_u_kbps  444771 non-null  int64  
 3   avg_lat_ms  444771 non-null  int64  
 4   tests       444771 non-null  int64  
 5   devices     444771 non-null  int64  
 6   type        444771 non-null  object 
 7   quarter     444771 non-null  object 
 8   year        444771 non-null  int64  
 9   STATEFP     444771 non-null  int64  
 10  COUNTYFP    444771 non-null  int64  
 11  TRACTCE     444771 non-null  int64  
 12  AFFGEOID    444771 non-null  object 
 13  GEOID       444771 non-null  int64  
 14  NAME        444771 non-null  float64
 15  LSAD        444771 non-null  object 
 16  ALAND       444771 non-null  int64  
 17  AWATER      444771 non-null  int64  
 18  tile        444771 non-null  object 
dtypes:

In [51]:
# get unique values of GEOID
ny_ookla.GEOID.nunique()

3891

In [52]:
# get unique values of tile column
ny_ookla.tile.nunique()

234446

In [53]:
# get unique values of quadkeys column
ny_ookla.quadkey.nunique()

234446

In [78]:
fixed_ookla_shp = gpd.read_file('data/z-archive/Ookla/ookla-data/2019-01-01_performance_fixed_tiles/gps_fixed_tiles.shp')
fixed_ookla_shp.head()

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices  \
0  0231113112003202       66216       12490          13     28        4   
1  1322111021111001      102598       37356          13     15        4   
2  3112203030003110       24686       18736          18    162      106   
3  0320000130321312       17674       13989          78    364        4   
4  0320001332313103      441192      218955          22     14        1   

                                            geometry  
0  POLYGON ((-90.65918 38.49229, -90.65369 38.492...  
1  POLYGON ((110.35217 21.28937, 110.35767 21.289...  
2  POLYGON ((138.59253 -34.92197, 138.59802 -34.9...  
3  POLYGON ((-87.63794 40.22502, -87.63245 40.225...  
4  POLYGON ((-84.74304 38.92096, -84.73755 38.920...

In [79]:
mobile_ookla_shp = gpd.read_file('data/z-archive/Ookla/ookla-data/2019-01-01_performance_mobile_tiles/gps_mobile_tiles.shp')
mobile_ookla_shp.head()

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices  \
0  1202130120303121       12503        5090          43      1        1   
1  0322113021201023       16109       11204          36     98       49   
2  3100130002212100       10325        7378          46     94       34   
3  1231213031333131        7187        2051          50    211      104   
4  0230102031111210       76282       20332          27      2        2   

                                            geometry  
0  POLYGON ((18.52844 51.91039, 18.53394 51.91039...  
1  POLYGON ((-69.91150 18.46398, -69.90601 18.463...  
2  POLYGON ((106.98486 -6.18425, 106.99036 -6.184...  
3  POLYGON ((77.33826 28.64239, 77.34375 28.64239...  
4  POLYGON ((-122.37671 38.25544, -122.37122 38.2...

```
intersects, contains, within, touches, crosses, overlap
```

`object.crosses(other)`
- Returns True if the interior of the object intersects the interior of the other but does not contain it, and the dimension of the intersection is less than the dimension of the one or the other. A line does not cross a point that it contains.

`object.contains(other)`
- Returns True if no points of other lie in the exterior of the object and at least one point of the interior of other lies in the interior of object. This predicate applies to all types, and is inverse to within(). The expression a.contains(b) == b.within(a) always evaluates to True. A line’s endpoints are part of its boundary and are therefore not contained.

`object.intersects(other)`
- Returns True if the boundary or interior of the object intersect in any way with those of the other. In other words, geometric objects intersect if they have any boundary or interior point in common.

In [81]:
us_tracts = gpd.read_file('data/z-archive/Census Tract Shapefiles/cb_2019_us_tract_500k/cb_2019_us_tract_500k.shp')
us_tracts.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID     NAME LSAD  \
0      34      005  702808  1400000US34005702808  34005702808  7028.08   CT   
1      39      153  504200  1400000US39153504200  39153504200     5042   CT   
2      36      071  000502  1400000US36071000502  36071000502     5.02   CT   
3      06      013  370000  1400000US06013370000  06013370000     3700   CT   
4      48      113  019204  1400000US48113019204  48113019204   192.04   CT   

     ALAND  AWATER                                           geometry  
0  1510553       0  POLYGON ((-74.90801 40.03250, -74.90496 40.036...  
1   808151       0  POLYGON ((-81.51807 41.05957, -81.51805 41.062...  
2   967431  969216  POLYGON ((-74.02226 41.49281, -74.02180 41.496...  
3   999356       0  POLYGON ((-122.32697 37.93502, -122.32610 37.9...  
4  2851229    9966  POLYGON ((-96.76902 32.94655, -96.76899 32.950...

In [82]:
ny_tracts = us_tracts[us_tracts['STATEFP']=='36']
ny_tracts.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID     NAME LSAD  \
2       36      071  000502  1400000US36071000502  36071000502     5.02   CT   
10      36      103  135208  1400000US36103135208  36103135208  1352.08   CT   
49      36      047  057800  1400000US36047057800  36047057800      578   CT   
50      36      047  058900  1400000US36047058900  36047058900      589   CT   
53      36      055  013204  1400000US36055013204  36055013204   132.04   CT   

       ALAND  AWATER                                           geometry  
2     967431  969216  POLYGON ((-74.02226 41.49281, -74.02180 41.496...  
10   2287077       0  POLYGON ((-73.28263 40.83063, -73.28157 40.832...  
49    172233       0  POLYGON ((-73.95398 40.60140, -73.95304 40.601...  
50    424025   38353  POLYGON ((-73.94605 40.72926, -73.94419 40.729...  
53  28207247   93541  POLYGON ((-77.66835 43.02829, -77.66806 43.029...

In [90]:
ny_tracts = ny_tracts.to_crs("EPSG:4326")

In [68]:
fl_tracts = us_tracts[us_tracts['STATEFP']=='12']
fl_tracts.head()

STATEFP COUNTYFP TRACTCE              AFFGEOID        GEOID     NAME LSAD  \
11      12      095  016902  1400000US12095016902  12095016902   169.02   CT   
12      12      095  016804  1400000US12095016804  12095016804   168.04   CT   
21      12      011  110311  1400000US12011110311  12011110311  1103.11   CT   
41      12      091  022001  1400000US12091022001  12091022001   220.01   CT   
58      12      011  060112  1400000US12011060112  12011060112   601.12   CT   

       ALAND  AWATER                                           geometry  
11   9854753  501139  POLYGON ((-81.44418 28.47207, -81.43981 28.472...  
12  16052327  165779  POLYGON ((-81.37578 28.41113, -81.37270 28.421...  
21   2366691  193908  POLYGON ((-80.28031 26.02264, -80.27249 26.022...  
41   1744725       0  POLYGON ((-86.63451 30.44949, -86.63219 30.449...  
58   2480149  142045  POLYGON ((-80.26840 26.20835, -80.25773 26.208...

In [92]:
fixed_tiles_ny = gpd.sjoin(fixed_ookla_shp, ny_tracts, how='inner', op='intersects')
fixed_ny = fixed_tiles_ny.reset_index(drop=True)
fixed_ny.head()

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices  \
0  0320101110331012      112600       79229           8     99       21   
1  0320101110331011      130626       67490           9    118       12   
2  0320101110331003      125738       95820           7     22       12   
3  0320101110331020       98630       65159          11     16        8   
4  0320101110313232      129067      111693           8     75       18   

                                            geometry  index_right STATEFP  \
0  POLYGON ((-73.50952 40.77638, -73.50403 40.776...        12888      36   
1  POLYGON ((-73.50403 40.78054, -73.49854 40.780...        12888      36   
2  POLYGON ((-73.51501 40.77638, -73.50952 40.776...        12888      36   
3  POLYGON ((-73.52051 40.77222, -73.51501 40.772...        12888      36   
4  POLYGON ((-73.50952 40.78470, -73.50403 40.784...        12888      36   

  COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD    ALAND  \
0      059  519400  1400000US36059519400  36059519400  5194   CT  1279432   
1      059  519400  1400000US36059519400  36059519400  5194   CT  1279432   
2      059  519400  1400000US36059519400  36059519400  5194   CT  1279432   
3      059  519400  1400000US36059519400  36059519400  5194   CT  1279432   
4      059  519400  1400000US36059519400  36059519400  5194   CT  1279432   

   AWATER  
0       0  
1       0  
2       0  
3       0  
4       0

In [93]:
mobile_tiles_ny = gpd.sjoin(mobile_ookla_shp, ny_tracts, how='inner', op='intersects')
mobile_ny = mobile_tiles_ny.reset_index(drop=True)
mobile_ny.head()

quadkey  avg_d_kbps  avg_u_kbps  avg_lat_ms  tests  devices  \
0  0302323301100220       25106        6284          56      1        1   
1  0302323301031300        3814        6047         109      2        2   
2  0302323301013330         913        7996          37      1        1   
3  0302323301013311       22793        5939          49      1        1   
4  0302323301031213        3581        5138          31      1        1   

                                            geometry  index_right STATEFP  \
0  POLYGON ((-74.00391 42.00849, -73.99841 42.008...        70893      36   
1  POLYGON ((-74.02588 41.95132, -74.02039 41.951...        70893      36   
2  POLYGON ((-74.01489 41.97583, -74.00940 41.975...        70893      36   
3  POLYGON ((-74.00940 41.98399, -74.00391 41.983...        70893      36   
4  POLYGON ((-74.03137 41.94723, -74.02588 41.947...        70893      36   

  COUNTYFP TRACTCE              AFFGEOID        GEOID  NAME LSAD     ALAND  \
0      111  951500  1400000US36111951500  36111951500  9515   CT  48023363   
1      111  951500  1400000US36111951500  36111951500  9515   CT  48023363   
2      111  951500  1400000US36111951500  36111951500  9515   CT  48023363   
3      111  951500  1400000US36111951500  36111951500  9515   CT  48023363   
4      111  951500  1400000US36111951500  36111951500  9515   CT  48023363   

   AWATER  
0  756736  
1  756736  
2  756736  
3  756736  
4  756736

In [94]:
fixed_ny.to_file('data/NY/fixed_ny.shp')
mobile_ny.to_file('data/NY/mobile_ny.shp')

/var/folders/0f/dj_hz1316w7c4qhdk33w3p3w0000gn/T/ipykernel_13796/4096207189.py:1: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.
  fixed_ny.to_file('data/NY/fixed_ny.shp')
/var/folders/0f/dj_hz1316w7c4qhdk33w3p3w0000gn/T/ipykernel_13796/4096207189.py:2: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.
  mobile_ny.to_file('data/NY/mobile_ny.shp')


In [95]:
# Check number of unique GEOID
len(mobile_ny.GEOID.value_counts())

4842

In [96]:
# Check number of unique GEOIDs
len(fixed_ny.GEOID.value_counts())

4894

I need to incorporate Group 1A's code for weighted averages, but I aggregated the data by taking the average or sum depending on the column.

In [97]:
wm = lambda x: np.average(x, weights=mobile_ny.loc[x.index, 'tests'])

mobile_agg = mobile_ny.groupby(['GEOID'], as_index=False).agg(
                            m_d_kbps_wm = ('avg_d_kbps', wm),
                            m_u_kbps_wm = ('avg_u_kbps', wm),
                            m_tests = ('tests', sum),
                            m_devices = ('devices', sum)
                            )
mobile_agg.head()

GEOID   m_d_kbps_wm   m_u_kbps_wm  m_tests  m_devices
0  36001000100  24171.750000  11767.375000       16         13
1  36001000200  22492.375000  11001.193182       88         33
2  36001000300  33222.559524  12156.107143       84         40
3  36001000401  51093.853659  14892.902439       41         20
4  36001000403  40498.808824  10517.750000       68         36

In [98]:
wm = lambda x: np.average(x, weights=fixed_ny.loc[x.index, 'tests'])

fixed_agg = fixed_ny.groupby(['GEOID'], as_index=False).agg(
                            f_d_kbps_wm = ('avg_d_kbps', wm),
                            f_u_kbps_wm = ('avg_u_kbps', wm),
                            f_tests = ('tests', sum),
                            f_devices = ('devices', sum)
                            )
fixed_agg.head()

GEOID    f_d_kbps_wm   f_u_kbps_wm  f_tests  f_devices
0  36001000100   65004.333333  10497.234568       81         28
1  36001000200   77159.277778  12049.351852      108         58
2  36001000300   84453.176923  22220.796154      260         95
3  36001000401   92439.055556  43385.986111       72         35
4  36001000403  102956.056980  13773.612536      351        115

In [99]:
fixed_agg['GEOID'] = fixed_agg['GEOID'].apply(lambda x: int(x))
mobile_agg['GEOID'] = mobile_agg['GEOID'].apply(lambda x: int(x))

In [100]:
df_ookla = mobile_agg.merge(fixed_agg, how='outer', on='GEOID')
df_ookla.head()

GEOID   m_d_kbps_wm   m_u_kbps_wm  m_tests  m_devices    f_d_kbps_wm  \
0  36001000100  24171.750000  11767.375000     16.0       13.0   65004.333333   
1  36001000200  22492.375000  11001.193182     88.0       33.0   77159.277778   
2  36001000300  33222.559524  12156.107143     84.0       40.0   84453.176923   
3  36001000401  51093.853659  14892.902439     41.0       20.0   92439.055556   
4  36001000403  40498.808824  10517.750000     68.0       36.0  102956.056980   

    f_u_kbps_wm  f_tests  f_devices  
0  10497.234568     81.0       28.0  
1  12049.351852    108.0       58.0  
2  22220.796154    260.0       95.0  
3  43385.986111     72.0       35.0  
4  13773.612536    351.0      115.0

In [101]:
df_ookla.shape

(4897, 9)

In [102]:
df_ookla['GEOID'] = df_ookla['GEOID'].apply(lambda x: str(x))
df_tot = joined_2.merge(df_ookla, how='left', on='GEOID')
df_tot.head()

GEOID                                           geometry  \
0  36001000100  POLYGON ((-73.74506 42.67228, -73.74374 42.677...   
1  36001000200  POLYGON ((-73.75993 42.65997, -73.76156 42.660...   
2  36001000300  POLYGON ((-73.83021 42.69418, -73.82785 42.696...   
3  36001000401  POLYGON ((-73.89421 42.70977, -73.89668 42.711...   
4  36001000403  POLYGON ((-73.82108 42.67857, -73.81916 42.680...   

   f_broadband  f_computer   f_ba  f_black  f_hispanic  log_mhi      mhi  \
0        0.836       0.896  0.137    0.740       0.128    10.39  32389.0   
1        0.767       0.903  0.225    0.845       0.043    10.23  27714.0   
2        0.819       0.881  0.289    0.435       0.121    10.72  45272.0   
3        0.837       0.857  0.476    0.076       0.046    11.22  74274.0   
4        0.817       0.861  0.591    0.129       0.058    11.22  74426.0   

   population  households    ba  den_ba    n_ba  nhh_computer  nhh_broadband  \
0      2035.0       800.0  13.7  1283.0   176.0           717            669   
1      4793.0      2213.0  22.5  2710.0   609.0          1999           1697   
2      6147.0      2362.0  28.9  3725.0  1078.0          2081           1935   
3      2362.0       997.0  47.6  2169.0  1032.0           854            834   
4      4253.0      2186.0  59.1  3220.0  1904.0          1882           1786   

   nhh_computer_any_internet  nhh_computer_and_dialup  \
0                        717                        0   
1                       1999                        0   
2                       2081                       11   
3                        854                        0   
4                       1882                        0   

   nhh_computer_and_broadband  nhh_computer_no_internet  nhh_no_computer  \
0                         669                        48               83   
1                        1685                       314              214   
2                        1917                       153              281   
3                         826                        28              143   
4                        1733                       149              304   

   n_children  n_children_computer  n_children_computer_and_dialup  \
0         532                  520                               0   
1        1274                 1221                               0   
2        1729                 1717                               0   
3         171                  171                               0   
4         462                  462                               0   

   n_children_computer_and_broadband  n_children_computer_no_internet  \
0                                511                                9   
1                               1138                               83   
2                               1630                               87   
3                                171                                0   
4                                462                                0   

   n_children_no_computer      max_dn     max_up      dn10     dn100  \
0                      12  711.250000  50.722222  1.875000  0.777778   
1                      53  940.000000  35.000000  2.529412  1.011765   
2                      12  928.040541  98.195946  2.290541  1.060811   
3                       0  928.975904  65.168675  2.156627  1.024096   
4                       0  940.000000  43.203883  2.310680  1.009709   

      dn250  fiber_100u  state   m_d_kbps_wm   m_u_kbps_wm  m_tests  \
0  0.777778    0.027778   36.0  24171.750000  11767.375000     16.0   
1  1.000000    0.000000   36.0  22492.375000  11001.193182     88.0   
2  1.060811    0.074324   36.0  33222.559524  12156.107143     84.0   
3  1.024096    0.036145   36.0  51093.853659  14892.902439     41.0   
4  1.009709    0.009709   36.0  40498.808824  10517.750000     68.0   

   m_devices    f_d_kbps_wm   f_u_kbps_wm  f_tests  f_devices  
0       13.0   65004.333333  10497.234568     81.0       28.0  
1       

In [103]:
df_tot.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Int64Index: 4906 entries, 0 to 4905
Data columns (total 42 columns):
 #   Column                             Non-Null Count  Dtype   
---  ------                             --------------  -----   
 0   GEOID                              4906 non-null   object  
 1   geometry                           4906 non-null   geometry
 2   f_broadband                        4906 non-null   float64 
 3   f_computer                         4906 non-null   float64 
 4   f_ba                               4906 non-null   float64 
 5   f_black                            4906 non-null   float64 
 6   f_hispanic                         4906 non-null   float64 
 7   log_mhi                            4785 non-null   float64 
 8   mhi                                4906 non-null   float64 
 9   population                         4906 non-null   float64 
 10  households                         4906 non-null   float64 
 11  ba                                 

In [104]:
df_tot.isna().sum()

GEOID                                  0
geometry                               0
f_broadband                            0
f_computer                             0
f_ba                                   0
f_black                                0
f_hispanic                             0
log_mhi                              121
mhi                                    0
population                             0
households                             0
ba                                     0
den_ba                                 0
n_ba                                   0
nhh_computer                           0
nhh_broadband                          0
nhh_computer_any_internet              0
nhh_computer_and_dialup                0
nhh_computer_and_broadband             0
nhh_computer_no_internet               0
nhh_no_computer                        0
n_children                             0
n_children_computer                    0
n_children_computer_and_dialup         0
n_children_compu

In [105]:
df_tot.to_file("data/NY/ny_combined.shp")

/var/folders/0f/dj_hz1316w7c4qhdk33w3p3w0000gn/T/ipykernel_13796/1179071031.py:1: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.
  df_tot.to_file("data/NY/ny_combined.shp")


# Illinois Datasets

# Formula to create dataset